opennlp-tools/src/main/java/opennlp/tools/ml/model/DataIndexerFactory.java

```diff
@@ -0,0 +1,66 @@
+package opennlp.tools.ml.model;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import opennlp.tools.ml.AbstractEventTrainer;
+import opennlp.tools.util.TrainingParameters;
+import opennlp.tools.util.ext.ExtensionLoader;
+
+public class DataIndexerFactory {
+  private static Map<String,Class<? extends DataIndexer>> _indexers = new HashMap<>();
+  static {
+    _indexers.put(AbstractEventTrainer.DATA_INDEXER_ONE_PASS_VALUE, OnePassDataIndexer.class);
+    _indexers.put(AbstractEventTrainer.DATA_INDEXER_TWO_PASS_VALUE, TwoPassDataIndexer.class);
+    _indexers.put("opennlp.tools.ml.model.OnePassRealValueDataIndexer", OnePassRealValueDataIndexer.class);
+  }
+  
+  /**
+   * Registers a new DataIndexer with the Factory.  The indexer class MUST have
+   * a noarg constructor.
+   * 
+   * @param key - A key for the trainingParameter.  Must be unique
+   * @param indexerClass - the class of the indexer returned.
+   * 
+   * @return true if the registration was successful, false if there was an error.
+   */
+  public static boolean registerDataIndexer(String key, Class<? extends DataIndexer> indexerClass) {
+    if (_indexers.containsKey(key)) return false;
+
+    _indexers.put(key, indexerClass);
+    return true;
+  }
+  
+  
+  public static DataIndexer getDataIndexer(TrainingParameters parameters,Map<String,String> reportMap) {
+    // The default is currently a 2-Pass data index.  Is this what we really want?
+    String indexerKey = parameters.getStringParameter(AbstractEventTrainer.DATA_INDEXER_PARAM, 
+        AbstractEventTrainer.DATA_INDEXER_TWO_PASS_VALUE);
+
+    // allow the user to pass in a report map.  If the don't, create one.
+    if (reportMap == null) reportMap = new HashMap<>();
+    
+    if (_indexers.containsKey(indexerKey)) {
```

Comment:
Probably the code looks much nicer if we just use a switch and new to instantiate those built-in ones.